docs/WritingTests.asciidoc

```diff
@@ -752,17 +768,169 @@ Sending key presses over VNC is very slow so for tests which send a lot of
+exists and can receive output. Some utility library functions are hard coded
+to redirect output to the emulated serial port and expect that +wait_serial+
+will be able to read it. Usually it is not too difficult to fix the utility
+function, you just need to remove some redirection in the shell command.
+
+Another common problem is that some library or utility function tries to take
+a screen shot. The hard part is finding what takes the screen shot, but then
+it is just a simple case of checking +is_serial_terminal+ and not taking it if
+we are on a serial terminal.
+
+As mentioned previously, distributions tend to wrap +select_console+, so
+instead of using it directly, you can use something like
+
+[source,perl]
+--------------------------------------------------------------------------------
+if (select_virtio_console()) {
+        # Do something which only works, or is necessary, on a serial terminal
+}
+--------------------------------------------------------------------------------
+
+This selects the virtio based serial terminal if possible. If it is available
+then it returns true. It is also possible to check if the current console is a
+serial terminal by calling +is_serial_terminal+.
+
+Once you have selected a serial terminal, the video feed will disappear from
+the live view, however at the bottom of the live screen there is a separate
+text feed. After the test has finished you can view the serial log(s) in the
+assets tab. You will probably have two serial logs; +serial0.txt+ and
++serial_terminal.txt+.
+
+Now that you are on a serial terminal everything will start to go a lot
+faster. So much faster in fact that race conditions become a big
+issue. Generally these can be avoided by using the higher level functions such
+as +script_run+ and +script_output+.
+
+It is rarely necessary to use the lower level functions, however it helps to
+recognise problems caused by race conditions at the lower level, so please
+read the following section regardless.
+
+So if you do need to use +type_string+ and +wait_serial+ directly then try to
+use the following pattern:
+
+1) Wait for the terminal prompt to appear.
+2) Send your command
+3) Wait for your command text to be echoed by the shell (inf applicable)
+4) Send enter
+5) Wait for your command output (if applicable)
+
+To illustrate this is a snippet from the LTP test runner which uses the lower
+level commands to achieve a little bit more control.
+
+[source,perl]
+--------------------------------------------------------------------------------
+my $fin_msg    = "### TEST $test->{name} COMPLETE >>> ";
+my $cmd_text   = qq($test->{command}; echo "$fin_msg\$?");
+my $klog_stamp = "echo 'OpenQA::run_ltp.pm: Starting $test->{name}' > /dev/$serialdev";
+
+# More variables and other shit
+
+if (is_serial_terminal) {
+        script_run($klog_stamp);
+        wait_serial(serial_term_prompt(), undef, 0, no_regex => 1);
```

Comment:
can you elaborate more about serial_term_prompt() ?

Reply:
actually it make sense to explain other parameters in wait_serial and why you using them. It will give some insights about how serial terminal actually working

Reply:
but maybe I am asking too much for single PR ... feel free to ignore this one

Reply:
Agree :-). I mean it's great to have comments but on the other hand perfect is the enemy of good :-).

Thank you both for your care.

Reply:
I think it is perfectly valid criticism :-)